hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockDataStreamOutput.java

```diff
@@ -438,12 +439,39 @@ private void executePutBlock(boolean close,
       byteBufferList = null;
     }
     waitFuturesComplete();
+    BlockData blockData = containerBlockData.build();
     if (close) {
-      dataStreamCloseReply = out.closeAsync();
+      out.writeAsync(ByteBuffer.allocateDirect(0).asReadOnlyBuffer())
```

Comment:
Writing an empty buffer may not work as expected.  In the streaming model, data is continuous but not block-by-block.  Both sender and receiver must consider that the data is a stream of bytes.

For example, the sender may write 3 times with lengths 3, 4, 5.  Since there could be buffers/encoder/decoder in between, the receiver may receive a single buffer with length 12.

Reply:
Thanks for your comment @szetszwo .   I wrote a piece of pseudo-code, Ideas about the modification of `org.apache.ratis.netty.server.DataStreamManagement#writeTo`.   

![image](https://user-images.githubusercontent.com/5122636/158544128-f60c7e94-efa3-412a-85ac-7304e284c241.png)

What do you think of this idea?

Reply:
@guohao-rosicky , this is a standard problem in streaming.  The standard solution is to write the length with the data.  Let's don't change the API for it.

Reply:
> 

Hi, @szetszwo I have a question,
I noticed that in HDFS, marking the end of stream is also represented by an empty package.
In Ratis Stream, sticky packets have already been dealt with.

org.apache.ratis.netty.NettyDataStreamUtils#encodeDataStreamRequestHeader
https://github.com/apache/ratis/blob/master/ratis-netty/src/main/java/org/apache/ratis/netty/NettyDataStreamUtils.java#L90